docs/inventory_plugins.md

```diff
@@ -0,0 +1,197 @@
+to plugins, provided that sufficient compatibility is in place for
+the version of Ansible present in the custom virtualenv where the import
+is being ran.
+
+To see what version the plugin transition will happen, see
+`awx/main/models/inventory.py` and look for the source name as a
+subclass of `PluginFileInjector`, and there should be an `initial_version`
+which is the first version that testing deemed to have sufficient parity
+in the content its inventory plugin returns. For example, `openstack` will
+begin using the inventory plugin in Ansible version 2.8, because the
+openstack inventory plugin had an issue with sending logs to stdout which
+was fixed in that version. If you run an openstack inventory update in
+2.8 or lower, it will use the script.
+
+### Sunsetting the scripts
+
+Eventually, it is intended that all source types will have moved to
+plugins. For any given source, after the `initial_version` for plugin use
+is higher than the lowest supported Ansible version, the script can be
+removed and the logic for script credential injection can also be removed.
+
+For example, after AWX no longer supports Ansible 2.7, the script
+`awx/plugins/openstack_inventory.py` will be removed.
+
+## Changes to Expect in Imports
+
+An effort was made to keep imports working in the exact same way after
+the switchover. However, the inventory plugins are a fundamental rewrite
+and many elements of default behavior has changed. Because of that,
+a `compatibility_mode` toggle was added.
+
+In a data migration, all existing cloud sources are switched over to
+use `compatibility_mode`. New inventory sources will default to having
+this off.
+
+We recommend that you opt out of compatibility mode because this is more
+future-proof, and also suggest that you set the `overwrite`
+flag to help assure stale content is removed.
+
+### Changes with Compatibility Mode Off
+
+If no `group_by` entries are given, then no constructed groups will be
+produced. That means no grouping by tags, regions, or similar attributes
+unless the user adds these to the `group_by` listing. This is different
+from prior behavior, where a blank `group_by` field would include all
+possible groups.
+
+The set of `hostvars` will be almost completely different, using new names
+for data which is mostly the same content. You can see the jinja2 keyed_groups
+construction used in compatibility mode to help get a sense of what
+new names replace old names.
+
+In many case, the host names will change. In many cases, accurate host
+tracking will still be maintained via the host `instance_id`, but this
+is not guaranteed.
+
+Group names will be sanitized. That means that characters such as "-" will
+be replaced by underscores "\_". In some cases, this means that a large
+fraction of groups get renamed as you move from scripts to plugins.
+Sanitizing group names allows referencing them in jinja2 in playbooks
+without errors.
+
```

Comment:
@AlanCoding I am unclear where we are landing with this. Are you doing some reverse sanitization in compat mode to get the `-`'s back? or we just hoping that lands

Reply:
You wanted to be able to run inventory imports with plugins with Ansible devel. Because of that, I turned on the big 3 for devel.

> Are you doing some reverse sanitization in compat mode to get the `-`'s back? or we just hoping that lands

We are counting on a fix for that landing. If a fix for that does not land, then I assume we will have to go in before the end of this development cycle and _disable_ the plugins for these sources.

However, until a fix actually lands, we will not know what the contract will be to use the unsafe group names (which we would only do in compat mode). Adding a parameter to each grouping, or adding an Ansible setting env var would be completely trivial, but since a toggle for this is still in limbo, we simply don't know what to do.

Reply:
Is ansible changing group names by default now?

Reply:
> Is ansible changing group names by default now?

No, the warning message states version 2.10 as the timeframe for that.